### PR TITLE
Metric aggregation

### LINKED ISSUE #3245: Metric Aggregation
## 🚀 Feature
To offer a better metric aggregation I discussed a potential way to go with @SkafteNicki . 

We agreed on the following:

* add a new `aggregated`-property that aggregates the metric over batches.
* default ddp-sync to broadcasting + aggregation
* add a `reset`  function to metrics to reset the internal persistent aggregation state (can be called by user whenever they want; usually on epoch end)

in pseudo code the metric workflow will now be:

```python
interm_outputs = do_some_first_calc(batch)
broadcasted_interm = broadcast(interm_outputs)
persistent_state.extend(broadcasted_interm)
return aggregate(broadcasted_interm)
```

cc @PyTorchLightning/core-contributors

### PR DESCRIPTION
**Pull request by justusschock**
Allows aggregation on the same device as well as across multiple devices in DDP

Fixes #3245 

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by justusschock** [comment]
@SkafteNicki Do we need a default reduce op for base classes? I don't think so, since currently it's all gathered and no op is done during ddp sync

File: pytorch_lightning/metrics/converters.py [hunk 1]
```diff
@@ -265,23 +279,34 @@ def sync_ddp_if_available(result: Union[torch.Tensor],
...
+        the optionally reshaped tensor
+    """
+    if tensor.shape == ():
+        tensor = tensor.reshape(1,)
+    return tensor
+
 
-def gather_all_tensors_if_available(result: Union[torch.Tensor],
-                                    group: Optional[Any] = None):
+def gather_all_tensors_if_available(result: Union[torch.Tensor], group: Optional[Any] = None):
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/converters.py:309]: Note that this function is has an error and we probably need https://github.com/PyTorchLightning/pytorch-lightning/pull/3319 merged first

File: pytorch_lightning/metrics/metric.py [hunk 2]
```diff
@@ -40,32 +44,41 @@ class Metric(DeviceDtypeModuleMixin, nn.Module, ABC):
...
         """
         super().__init__()
         self.name = name
         self._dtype = torch.get_default_dtype()
-        self._device = torch.device('cpu')
+        self._device = torch.device("cpu")
+
+        self.reduce_group = reduce_group
+
+        self._step_vals = []
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:76]: Is it possible for us to make this a buffer, so saving the model and thereby the metric will also save the current internal state of the metric

File: pytorch_lightning/metrics/metric.py [hunk 3]
```diff
@@ -151,6 +181,15 @@ def compute(self, data: Any, output: Any):
         """
         return output
 
+    @property
+    def aggregated(self) -> torch.Tensor:
+        aggr = self.aggregate(*self._step_vals)
+        self.reset()
+        return self.compute(self, None, aggr)
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:188]: Very nice :]

File: tests/metrics/test_metrics.py [hunk 4]
```diff
@@ -186,3 +193,93 @@ def test_saving_pickable(tmpdir, metric: Metric):
...
+        "aggregated",
+        "aggregate",
+        "reset",
+        "input_convert",
+        "forward",
+        "output_convert",
+        "ddp_reduce",
+        "ddp_sync",
+        "aggregate",
+    ]
```
Comment (SkafteNicki) [inline tests/metrics/test_metrics.py:285]: Could we add a third test: reinit metric to reset call history, two calls to the metric and then agggregate (case that most corresponds to trying to accumulate over batches

**Review by SkafteNicki — COMMENTED** [review]
Really nice changes:
I completely agree that we do not need the `reduce_op` argument anymore for the base class.

File: pytorch_lightning/metrics/metric.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py:76]: Actually this will be difficult, since buffers can only be tensors and tensors have a fixed size. Also buffers are moved to the device accordingly. We probably don't want this, but want it to be on cpu for the time of storing and want to move it back to the device after aggregation.

On the other hand, I'm not sure if it won't be saved anyway, since pytorch uses basically pickle for saving.

File: tests/metrics/test_metrics.py (hunk 4 above)
Comment (justusschock) [reply tests/metrics/test_metrics.py:285]: sure, will add that one :)

File: pytorch_lightning/metrics/metric.py (hunk 2 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:76]: Alright, good point.
Even IF they are not saved I don't see it as a huge drawback, more in sense that it would be nice to have

File: pytorch_lightning/metrics/converters.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/metrics/converters.py:309]: I saw that one, but I decided to use this for implementation anyways, since I knew it will be fixed :)

File: tests/metrics/test_metrics.py (hunk 4 above)
Comment (justusschock) [reply tests/metrics/test_metrics.py:285]: done :)

**Comment by justusschock** [comment]
@SkafteNicki I fixed the tests to be compliant to our current behaviour. All changes to aggregation will be done in following PRs :)

File: pytorch_lightning/metrics/classification.py [hunk 5]
```diff
@@ -106,20 +102,20 @@ class ConfusionMatrix(TensorMetric):
...
-            reduce_op: Any = None,
+        self,
+        normalize: bool = False,
+        reduce_group: Any = None,
     ):
         """
         Args:
             normalize: whether to compute a normalized confusion matrix
             reduce_group: the process group to reduce metric results from DDP
             reduce_op: the operation to perform for ddp reduction
```
Comment (awaelchli) [inline pytorch_lightning/metrics/classification.py:113]: reduce_op got removed, doc string still has it
same in all other places where reduce_op got removed

File: pytorch_lightning/metrics/classification.py [hunk 6]
```diff
@@ -133,8 +129,15 @@ def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
         Return:
             A Tensor with the confusion matrix.
         """
-        return confusion_matrix(pred=pred, target=target,
-                                normalize=self.normalize)
+        return confusion_matrix(pred=pred, target=target, normalize=self.normalize)
+
+    def aggregate(self, *tensors: torch.Tensor) -> torch.Tensor:
+        """Aggregates results by stacking them instead of concatenating before averaging.
```
Comment (awaelchli) [inline pytorch_lightning/metrics/classification.py:135]: only works when all tensors have the same shape, but that's probably the case, right?

File: pytorch_lightning/metrics/converters.py [hunk 7]
```diff
@@ -265,23 +278,37 @@ def sync_ddp_if_available(result: Union[torch.Tensor],
...
+    Args:
+        tensor: the tensor or array to check the shape for
+
+    Returns:
+        the optionally reshaped tensor
+    """
+    if tensor.shape == ():
+        tensor = tensor.reshape(
+            1,
+        )
```
Comment (awaelchli) [inline pytorch_lightning/metrics/converters.py:307]: did black do that? xD

File: pytorch_lightning/metrics/metric.py [hunk 8]
```diff
@@ -159,38 +197,17 @@ class TensorMetric(Metric):
...
     @staticmethod
     def input_convert(self, data: Any):
-        return apply_to_collection(data,
-                                   (torch.Tensor, np.ndarray, numbers.Number),
-                                   convert_to_tensor,
-                                   self.dtype, self.device)
+        data = apply_to_collection(
+            data, (torch.Tensor, np.ndarray, numbers.Number), convert_to_tensor, self.dtype, self.device
+        )
+        return super(TensorMetric, self).input_convert(self, data)
```
Comment (awaelchli) [inline pytorch_lightning/metrics/metric.py:205]: Is it important to name the class in the super call?

File: pytorch_lightning/metrics/sklearns.py [hunk 9]
```diff
@@ -1103,9 +1032,16 @@ def forward(
...
+        return np.array(super().forward(probas_pred=probas_pred, y_true=y_true, sample_weight=sample_weight)[:2])
+
+    def aggregate(self, *tensors: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Aggregates results by stacking them instead of concatenating before averaging.
+
+        Returns:
+            the aggregated results
+        """
+        print(tensors)
+        return tuple([torch.stack(tmp).mean(0) for tmp in zip(*tensors)])
```
Comment (awaelchli) [inline pytorch_lightning/metrics/sklearns.py:1044]:
```suggestion
        """
        return tuple([torch.stack(tmp).mean(0) for tmp in zip(*tensors)])
```
save the ink, save the trees!

**Review by awaelchli — APPROVED** [review]
a bit much with all the black formatting :)
just two feedbacks:
1. reduce op docstring is left over
2. stacking only works if all the dims are the same, is this assumption correct?

File: tests/metrics/test_sklearn.py [hunk 10]
```diff
@@ -167,13 +167,15 @@ def test_sklearn_metric(metric_class, sklearn_func, inputs):
 
     sklearn_result = sklearn_func(**numpy_inputs)
     lightning_result = metric_class(**inputs)
-    assert np.allclose(sklearn_result, lightning_result, atol=1e-5)
+
+    
+    # assert np.allclose(sklearn_result, lightning_result, atol=1e-5)
```
Comment (awaelchli) [inline tests/metrics/test_sklearn.py:172]:
```suggestion
```

File: pytorch_lightning/metrics/classification.py (hunk 5 above)
Comment (justusschock) [reply pytorch_lightning/metrics/classification.py:113]: Will remove that tomorrow

File: pytorch_lightning/metrics/classification.py (hunk 6 above)
Comment (justusschock) [reply pytorch_lightning/metrics/classification.py:135]: That is an assumption we make and for almost every case it holds. If it doesn't, you have to overwrite this method.

File: pytorch_lightning/metrics/converters.py (hunk 7 above)
Comment (justusschock) [reply pytorch_lightning/metrics/converters.py:307]: Yep :D

File: pytorch_lightning/metrics/metric.py (hunk 8 above)
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py:205]: It is, since this is a static method and thus the super call cannot infer the base class otherwise.

Unfortunately we also can't make this a non-static method since PyTorch requires all it's hooks to get the module passed explicitly

Comment (awaelchli) [reply pytorch_lightning/metrics/metric.py:205]: got it!

File: pytorch_lightning/metrics/converters.py [hunk 11]
```diff
@@ -31,10 +32,20 @@
...
+    rank_zero_warn("Unsupported `ReduceOp` for distributed computing")
+
+try:
+    import torch_xla
+    import torch_xla.core.xla_model as xm
+    import torch_xla.core.functions as xf
+except ImportError:
+    XLA_AVAILABLE = False
+else:
+    XLA_AVAILABLE = True
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/converters.py:48]: Is this used anywhere?

File: pytorch_lightning/metrics/metric.py [hunk 12]
```diff
@@ -159,38 +197,17 @@ class TensorMetric(Metric):
...
-
-    @staticmethod
-    def ddp_sync(self, data: Any, output: Any):
-        return apply_to_collection(output, torch.Tensor, sync_ddp_if_available,
-                                   self.reduce_group, self.reduce_op)
+        output = apply_to_collection(output, torch.Tensor, convert_to_tensor, self.dtype, self.device)
+        return super(TensorMetric, self).output_convert(self, data, output)
 
 
 class TensorCollectionMetric(Metric):
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:213]: We should at some point figure out to merge `TensorMetric` with `TensorCollectionMetric` since the difference seems very subtle

File: tests/metrics/test_metrics.py [hunk 13]
```diff
@@ -186,3 +188,120 @@ def test_saving_pickable(tmpdir, metric: Metric):
...
+
+        def ddp_sync(self, tensor: Any):
+            self.call_history.append("ddp_sync")
+            return super().ddp_sync(tensor)
+
+        @staticmethod
+        def ddp_reduce(self, data: Any, output: Any):
+            self.call_history.append("ddp_reduce")
+            return super(DummyMetric, self).ddp_reduce(self, data, output)
+
```
Comment (SkafteNicki) [inline tests/metrics/test_metrics.py:221]: We need to add the `compute` method here (not that we use that yet)

**Review by SkafteNicki — APPROVED** [review]
Nice work 👍 only few smaller comments

File: pytorch_lightning/metrics/converters.py (hunk 11 above)
Comment (justusschock) [reply pytorch_lightning/metrics/converters.py:48]: Ah no. I was working on a function to also do the same on TPU, but for now, I won't include this here

File: pytorch_lightning/metrics/metric.py (hunk 12 above)
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py:213]: we could probably do a try on tensor aggregation and catch it for TensorCollection. And same for output convert

Comment (justusschock) [reply pytorch_lightning/metrics/metric.py]: Think I already merged these

Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py]: Just something to think about, it would make it a bit easier for the user to implement own stuff

**Comment by SkafteNicki** [comment]
we have one test that are failing: `test_model_pickable[DummyNumpyMetric]` in file `tests/metrics/test_metrics`. 
The test succeeds when we are using `DummyTensorMetric` but fails with `DummyNumpyMetric`. I wonder if the error is fixed, if the the special aggregation for `TensorMetric` also is implemented for `NumpyMetric`?

File: pytorch_lightning/metrics/metric.py [hunk 14]
```diff
@@ -151,6 +180,15 @@ def compute(self, data: Any, output: Any):
         """
         return output
 
+    @property
+    def aggregated(self) -> torch.Tensor:
+        aggr = self.aggregate(*self._step_vals)
+        self.reset()
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/metrics/metric.py:186]: @SkafteNicki ?

[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:186]: It should reset `self._step_vals`variable such that the buffer is emptied after each call to aggregated property

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
> we have one test that are failing: `test_model_pickable[DummyNumpyMetric]` in file `tests/metrics/test_metrics`.
> The test succeeds when we are using `DummyTensorMetric` but fails with `DummyNumpyMetric`. I wonder if the error is fixed, if the the special aggregation for `TensorMetric` also is implemented for `NumpyMetric`?

it seems to be appearing just in this PR, it is not presented in master... it shall be patched so we do not break master...

**Review by Borda — CHANGES_REQUESTED** [review]
let's fix the failing test...

**Comment by s-rog** [comment]
Are there any plans to support variable sizes (in batch dimension) in gather all?

**Comment by SkafteNicki** [comment]
@Borda I fixed the failing test, can I merge this?
